pkg/graveler/graveler.go

```diff
@@ -1606,29 +1614,59 @@ func (g *KVGraveler) List(ctx context.Context, repositoryID RepositoryID, ref Re
 }
 
 func (g *KVGraveler) Commit(ctx context.Context, repositoryID RepositoryID, branchID BranchID, params CommitParams) (CommitID, error) {
-	// TODO(issue 3569) - kv implementation optimistic lock commit
 	var preRunID string
 	var commit Commit
+	var newCommitID CommitID
 	var storageNamespace StorageNamespace
-	res, err := g.db.branchLocker.MetadataUpdater(ctx, repositoryID, branchID, func() (interface{}, error) {
-		isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_COMMIT)
+	var sealedToDrop []StagingToken
+
+	isProtected, err := g.protectedBranchesManager.IsBlocked(ctx, repositoryID, branchID, BranchProtectionBlockedAction_COMMIT)
+	if err != nil {
+		return "", err
+	}
+	if isProtected {
+		return "", ErrCommitToProtectedBranch
+	}
+	repo, err := g.RefManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return "", fmt.Errorf("get repository: %w", err)
+	}
+	storageNamespace = repo.StorageNamespace
+
+	branch, err := g.GetBranch(ctx, repositoryID, branchID)
+	if err != nil {
+		return "", fmt.Errorf("get branch: %w", err)
+	}
+
+	if params.SourceMetaRange != nil {
+		empty, err := g.stagingEmpty(ctx, repositoryID, repo, branch)
 		if err != nil {
-			return nil, err
-		}
-		if isProtected {
-			return nil, ErrCommitToProtectedBranch
+			return "", fmt.Errorf("checking empty branch: %w", err)
 		}
-		repo, err := g.GetRepository(ctx, repositoryID)
-		if err != nil {
-			return "", fmt.Errorf("get repository: %w", err)
+		if !empty {
+			return "", ErrCommitMetaRangeDirtyBranch
 		}
-		storageNamespace = repo.StorageNamespace
+	}
 
-		branch, err := g.db.GetBranch(ctx, repositoryID, branchID)
-		if err != nil {
-			return "", fmt.Errorf("get branch: %w", err)
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(branch *Branch) (*Branch, error) {
+		if params.SourceMetaRange != nil {
```

Comment:
I think you also need to check for empty changes for regular commit

Reply:
It is checked by commit operation on the commit manager.
This check is for non-empty changes for the source meta range.

Reply:
I also added another test to check it

Reply:
Still - I think we should be avoiding adding another staging token the commit cannot succeed. Adding a staging token increases the latency of all read operations.
I think it should look something like:
```suggestion
			empty, err := g.stagingEmpty(ctx, repositoryID, repo, branch)
			if err != nil {
				return nil, fmt.Errorf("checking empty branch: %w", err)
			}
			if params.SourceMetaRange == nil && empty{
				return nil, ErrTypeThatWeUseForThis
			}
			if params.SourceMetaRange != nil && !empty {
				return nil, ErrCommitMetaRangeDirtyBranch
			}
```